File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/FileDataService.java
package sk.styk.martin.apkanalyzer.business.service;

import android.content.pm.PackageInfo;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.jar.Attributes;
import java.util.jar.JarFile;
import java.util.jar.Manifest;

import sk.styk.martin.apkanalyzer.model.detail.FileData;
import sk.styk.martin.apkanalyzer.model.detail.FileEntry;

/**
 * Created by Martin Styk on 30.06.2017.
 */
public class FileDataService {

    @NonNull
    public FileData get(@NonNull PackageInfo packageInfo) {

        FileData fileData = new FileData();

        Manifest mf = openManifest(packageInfo);

        if (mf == null) {
            return fileData;
        }

        List<FileEntry> drawables = new ArrayList<>();
        List<FileEntry> layouts = new ArrayList<>();
        List<FileEntry> menus = new ArrayList<>();
        List<FileEntry> others = new ArrayList<>();

        int numberPngs = 0, numberXmls = 0, numberPngsWithDifferentName = 0, numberXmlsWithDifferentName = 0;
        Set<String> pngsSet = new HashSet<>();
        Set<String> xmlsSet = new HashSet<>();

        Map<String, Attributes> map = mf.getEntries();

        for (Map.Entry<String, Attributes> entry : map.entrySet()) {
            String fileName = entry.getKey();
            String hash = extractHash(entry);
            FileEntry fileEntry = new FileEntry(fileName, hash);

            // sort into categories according to location
            if (fileName.startsWith("res/drawable")) {
                drawables.add(fileEntry);

            } else if (fileName.startsWith("res/layout")) {
                layouts.add(fileEntry);

            } else if (fileName.startsWith("res/menu")) {
                menus.add(fileEntry);

            } else if (fileName.equals("classes.dex")) {
                fileData.setDexHash(hash);

            } else if (fileName.equals("resources.arsc")) {
                fileData.setArscHash(hash);

            } else if (fileName.equals("AndroidManifest.xml")) {
                fileData.setManifestHash(hash);

            } else {
                others.add(fileEntry);
            }

            // sort into categories according to file type
            if (fileName.endsWith(".png")) {
                numberPngs++;
                pngsSet.add(fileEntry.getFileName());
            } else if (fileName.endsWith(".xml")) {
                numberXmls++;
                xmlsSet.add(fileEntry.getFileName());
            }

        }

        fileData.setDrawableHashes(drawables);
        fileData.setLayoutHashes(layouts);
        fileData.setMenuHashes(menus);
        fileData.setOtherHashes(others);

        fileData.setNumberPngs(numberPngs);
        fileData.setNumberPngsWithDifferentName(pngsSet.size());
        fileData.setNumberXmls(numberXmls);
        fileData.setNumberXmlsWithDifferentName(xmlsSet.size());

        return fileData;
    }

    /**
     * Gets all entries from application manifest file
     *
     * @param packageInfo APK packageInfo
     * @return all entries on manifest file or null, manifest is not found
     */
    @Nullable
    private Manifest openManifest(PackageInfo packageInfo) {
        if(packageInfo.applicationInfo == null || packageInfo.applicationInfo.sourceDir == null)
            return null;

        Manifest mf = null;
        try {
            JarFile jar = new JarFile(packageInfo.applicationInfo.sourceDir);
            mf = jar.getManifest();
        } catch (IOException e) {
            Log.e(FileDataService.class.getSimpleName(), "Unable to find manifest", e);
        }
        return mf;
    }

    private String extractHash(Map.Entry<String, Attributes> entry) {
        String sha1 = (entry.getValue() == null) ? null : entry.getValue().getValue("SHA1-Digest");
        if (sha1 != null)
            return sha1;

        String sha256 = (entry.getValue() == null) ? null : entry.getValue().getValue("SHA-256-Digest");
        return sha256;
    }
}

File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/FileDataService.kt
package sk.styk.martin.apkanalyzer.business.service

import android.content.pm.PackageInfo
import android.util.Log
import sk.styk.martin.apkanalyzer.model.detail.FileData
import sk.styk.martin.apkanalyzer.model.detail.FileEntry
import java.io.IOException
import java.util.*
import java.util.jar.Attributes
import java.util.jar.JarFile
import java.util.jar.Manifest
import kotlin.collections.Map.Entry

/**
 * @author Martin Styk
 * @version 30.06.2017.
 */
class FileDataService {

    fun get(packageInfo: PackageInfo): FileData {

        val fileData = FileData()

        val mf = openManifest(packageInfo) ?: return fileData

        val drawables = ArrayList<FileEntry>()
        val layouts = ArrayList<FileEntry>()
        val menus = ArrayList<FileEntry>()
        val others = ArrayList<FileEntry>()

        var numberPngs = 0
        var numberXmls = 0
        val pngsSet = HashSet<String>()
        val xmlsSet = HashSet<String>()

        val map = mf.entries

        for (entry in map.entries) {
            val fileName = entry.key
            val hash = extractHash(entry)
            val fileEntry = FileEntry(fileName, hash)

            // sort into categories according to location
            when {
                fileName.startsWith("res/drawable") -> drawables.add(fileEntry)
                fileName.startsWith("res/layout") -> layouts.add(fileEntry)
                fileName.startsWith("res/menu") -> menus.add(fileEntry)
                "classes.dex".equals(fileName) -> fileData.dexHash = hash
                "resources.arsc".equals(fileName) -> fileData.arscHash = hash
                "AndroidManifest.xml".equals(fileName) -> fileData.manifestHash = hash
                else -> others.add(fileEntry)
            }

            // sort into categories according to file type
            when {
                fileName.endsWith(".png") -> {
                    numberPngs++
                    pngsSet.add(fileName)
                }
                fileName.endsWith(".xml") -> {
                    numberXmls++
                    xmlsSet.add(fileEntry.fileName)
                }
            }
        }

        fileData.drawableHashes = drawables
        fileData.layoutHashes = layouts
        fileData.menuHashes = menus
        fileData.otherHashes = others

        fileData.numberPngs = numberPngs
        fileData.numberPngsWithDifferentName = pngsSet.size
        fileData.numberXmls = numberXmls
        fileData.numberXmlsWithDifferentName = xmlsSet.size

        return fileData
    }

    /**
     * Gets all entries from application manifest file
     *
     * @param packageInfo APK packageInfo
     * @return all entries on manifest file or null, manifest is not found
     */
    private fun openManifest(packageInfo: PackageInfo): Manifest? {
        if (packageInfo.applicationInfo == null || packageInfo.applicationInfo.sourceDir == null)
            return null

        var jar: JarFile? = null
        try {
            jar = JarFile(packageInfo.applicationInfo.sourceDir)
            return jar.manifest
        } catch (e: IOException) {
            Log.e(FileDataService::class.java.simpleName, "Unable to find manifest", e)
        } finally {
            try {
                jar?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        return null
    }

    private fun extractHash(entry: Entry<String, Attributes>): String {
        return entry.value.getValue("SHA1-Digest") ?: entry.value.getValue("SHA-256-Digest")
    }
}
